File: src/verify.ts
/**
 * verifyProof — the one server-side entry point. Chain-agnostic: parse the
 * CAIP-122 message, enforce domain/nonce/time, then dispatch to the per-chain
 * cryptographic verifier. Fails closed: any unknown scheme or malformed input
 * returns `{ ok: false, reason }`, never throws.
 *
 * This pure function is mirrored by the Go port in go/walletconnect so IAM
 * verifies identically.
 *
 * Two entry points, same checks:
 *   - {@link verifyProof}      — synchronous, covers the five @noble-pure
 *     chains (evm/solana/bitcoin/ton/xrp). Pulls no WASM.
 *   - {@link verifyProofAsync} — covers ALL chains including Polkadot (whose
 *     sr25519 verify needs `cryptoWaitReady()`) and Cardano (CIP-8 COSE_Sign1).
 *     It delegates the five sync chains to the same code and lazy-loads the
 *     Substrate / Cardano verifiers for the rest.
 */
import type { SignedProof, VerifyExpectation, VerifyResult, Chain } from './types.js';
import { parseSiwxMessage } from './caip122.js';
import { verifyEvm } from './evm/verify.js';
import { verifySolana } from './solana/verify.js';
import { verifyTon } from './ton/verify.js';
import { verifyBitcoin } from './bitcoin/verify.js';
import { verifyXrp } from './xrp/verify.js';

const DEFAULT_SKEW_MS = 5 * 60 * 1000;

function fail(reason: NonNullable<VerifyResult['reason']>): VerifyResult {
  return { ok: false, reason };
}

/** Case-insensitive only for EVM (checksummed hex); all others are exact. */
function addressesEqual(chain: Chain, a: string, b: string): boolean {
  const x = a.trim();
  const y = b.trim();
  return chain === 'evm' ? x.toLowerCase() === y.toLowerCase() : x === y;
}

function parseTime(s: string | undefined): number | null {
  if (s == null) return null;
  const t = Date.parse(s);
  return Number.isNaN(t) ? null : t;
}

/** Synchronous cryptographic dispatch. Returns null for schemes this path does
 * not handle (Substrate, which needs the async verifier). */
function verifyCrypto(proof: SignedProof): boolean | null {
  switch (proof.scheme) {
    case 'secp256k1-eip191':
      return verifyEvm(proof.message, proof.signature, proof.address);
    case 'ed25519':
      // ed25519-over-message is Solana today; TON uses 'ton-proof'.
      return verifySolana(proof.message, proof.signature, proof.address);
    case 'ton-proof':
      return verifyTon(proof);
    case 'bip322':
      return verifyBitcoin(proof);
    case 'secp256k1-xrpl':
    case 'ed25519-xrpl':
      return verifyXrp(proof);
    default:
      return null;
  }
}

/**
 * Run every check up to (but excluding) the cryptographic one. Returns the
 * failing {@link VerifyResult} when a binding/time check fails, or `null` when
 * all of them pass (so the caller proceeds to crypto). Shared by both the sync
 * and async entry points so the order and reasons are identical.
 */
function preCrypto(proof: SignedProof, expected: VerifyExpectation): VerifyResult | null {
  let parsed;
  try {
    parsed = parseSiwxMessage(proof.message);
  } catch {
    return fail('malformed-message');
  }

  // 1. Binding: the signer in the message must match the proof's address.
  if (!addressesEqual(proof.chain, parsed.address, proof.address)) {
    return fail('address-mismatch');
  }
  if (expected.address != null && !addressesEqual(proof.chain, proof.address, expected.address)) {
    return fail('address-mismatch');
  }

  // 2. Domain + nonce binding (anti-phishing, anti-replay).
  if (parsed.domain !== expected.domain) {
    return fail('domain-mismatch');
  }
  if (parsed.nonce !== expected.nonce) {
    return fail('nonce-mismatch');
  }

  // 3. Time window.
  const now = expected.now ?? Date.now();
  const skew = expected.clockSkewMs ?? DEFAULT_SKEW_MS;
  const exp = parseTime(parsed.expirationTime);
  if (exp != null && now > exp + skew) {
    return fail('expired');
  }
  const nbf = parseTime(parsed.notBefore);
  if (nbf != null && now + skew < nbf) {
    return fail('not-yet-valid');
  }
  const iat = parseTime(parsed.issuedAt);
  if (iat != null && iat - skew > now) {
    return fail('not-yet-valid');
  }

  return null;
}

/** Map a crypto result (`true`/`false`) to the final {@link VerifyResult}. */
function finish(proof: SignedProof, ok: boolean): VerifyResult {
  if (!ok) return fail('bad-signature');
  return { ok: true, address: proof.address, chain: proof.chain };
}

/**
 * Verify a {@link SignedProof} synchronously. Handles the five @noble-pure
 * chains (evm/solana/bitcoin/ton/xrp). For Polkadot (Substrate sr25519/ed25519/
 * ecdsa) use {@link verifyProofAsync} — a Substrate proof here returns
 * `unsupported-scheme`.
 */
export function verifyProof(proof: SignedProof, expected: VerifyExpectation): VerifyResult {
  const pre = preCrypto(proof, expected);
  if (pre != null) return pre;

  const crypto = verifyCrypto(proof);
  if (crypto == null) {
    return fail('unsupported-scheme');
  }
  return finish(proof, crypto);
}

/**
 * Verify a {@link SignedProof} for ANY supported chain, including Polkadot.
 * Identical binding/time checks and reasons as {@link verifyProof}; only the
 * cryptographic step is async (Substrate sr25519 needs `cryptoWaitReady()`).
 * The five sync chains are delegated to the same code — no behavioural drift.
 */
export async function verifyProofAsync(
  proof: SignedProof,
  expected: VerifyExpectation,
): Promise<VerifyResult> {
  const pre = preCrypto(proof, expected);
  if (pre != null) return pre;

  switch (proof.scheme) {
    case 'sr25519':
    case 'ed25519-substrate':
    case 'ecdsa-substrate': {
      const { verifyPolkadot } = await import('./polkadot/verify.js');
      return finish(proof, await verifyPolkadot(proof));
    }
    case 'ed25519-cardano': {
      // Cardano's verifier is pure-synchronous (@noble + inline CBOR/bech32, no
      // WASM), but it is lazy-loaded here so the sync verify core never pulls
      // the CBOR/bech32 modules. Sits on the async path with Polkadot.
      const { verifyCardano } = await import('./cardano/verify.js');
      return finish(proof, verifyCardano(proof));
    }
    default: {
      const crypto = verifyCrypto(proof);
      if (crypto == null) return fail('unsupported-scheme');
      return finish(proof, crypto);
    }
  }
}

File: go/walletconnect/verify.go
package walletconnect

import (
	"strings"
	"time"
)

// SignatureScheme is the signing scheme that produced a proof. The verifier
// dispatches on this, not on Chain. Mirrors the TS SignatureScheme union.
type SignatureScheme string

const (
	SchemeSecp256k1EIP191 SignatureScheme = "secp256k1-eip191" // EVM personal_sign (EIP-191)
	SchemeEd25519         SignatureScheme = "ed25519"          // Solana, TON
	SchemeBIP322          SignatureScheme = "bip322"           // Bitcoin message signing (BIP-322)
	SchemeTonProof        SignatureScheme = "ton-proof"        // TON Connect ton_proof (ed25519 inside)
	SchemeSecp256k1XRPL   SignatureScheme = "secp256k1-xrpl"   // XRPL signMessage (secp256k1)
	SchemeEd25519XRPL     SignatureScheme = "ed25519-xrpl"     // XRPL ed25519 keypair
)

// Proof is what a wallet hands back after signing — everything the server needs
// to verify. Mirrors the TS SignedProof.
type Proof struct {
	Chain     Chain
	Scheme    SignatureScheme
	Address   string
	PublicKey string
	Message   string
	Signature string
	Extra     map[string]any
}

// Expectation holds the server-side expectations checked against the parsed
// message during verify. Mirrors the TS VerifyExpectation, but expressed with
// Go-native time types: Now overrides "now" (zero value => time.Now()) and
// ClockSkew is the tolerated skew (zero value => DefaultClockSkew).
type Expectation struct {
	Domain    string
	Nonce     string
	Address   string // optional; require an exact address (case-insensitive for EVM)
	Now       time.Time
	ClockSkew time.Duration
}

// DefaultClockSkew matches the TS DEFAULT_SKEW_MS (5 minutes).
const DefaultClockSkew = 5 * time.Minute

// Reason is a machine-readable failure reason. The string values match the TS
// VerifyResult.reason union exactly.
type Reason string

const (
	ReasonBadSignature      Reason = "bad-signature"
	ReasonAddressMismatch   Reason = "address-mismatch"
	ReasonDomainMismatch    Reason = "domain-mismatch"
	ReasonNonceMismatch     Reason = "nonce-mismatch"
	ReasonExpired           Reason = "expired"
	ReasonNotYetValid       Reason = "not-yet-valid"
	ReasonMalformedMessage  Reason = "malformed-message"
	ReasonUnsupportedScheme Reason = "unsupported-scheme"
	ReasonMissingPublicKey  Reason = "missing-public-key"
)

// Result mirrors the TS VerifyResult. Reason is empty when OK is true.
type Result struct {
	OK      bool
	Reason  Reason
	Address string
	Chain   Chain
}

func fail(r Reason) Result { return Result{OK: false, Reason: r} }

// addressesEqual is case-insensitive only for EVM (checksummed hex); all other
// chains compare exactly. Mirrors the TS addressesEqual.
func addressesEqual(chain Chain, a, b string) bool {
	x := strings.TrimSpace(a)
	y := strings.TrimSpace(b)
	if chain == ChainEVM {
		return strings.EqualFold(x, y)
	}
	return x == y
}

// parseTime parses an ISO-8601 timestamp the way JS Date.parse does for the
// values newChallenge emits (RFC3339 with milliseconds and a 'Z' zone). It
// tolerates the RFC3339 variants Date.parse accepts. ok=false means the field
// was absent or unparseable — both treated as "no constraint", matching the TS
// `parseTime` returning null.
func parseTime(s *string) (time.Time, bool) {
	if s == nil {
		return time.Time{}, false
	}
	for _, layout := range []string{time.RFC3339Nano, time.RFC3339} {
		if t, err := time.Parse(layout, *s); err == nil {
			return t, true
		}
	}
	return time.Time{}, false
}

// crypto dispatches to the per-chain cryptographic verifier. The (ok, supported)
// pair mirrors the TS `boolean | null`: supported=false maps to TS null
// (unsupported scheme); supported=true with ok=false maps to TS false.
func crypto(p Proof) (ok, supported bool) {
	switch p.Scheme {
	case SchemeSecp256k1EIP191:
		return VerifyEVM(p.Message, p.Signature, p.Address), true
	case SchemeEd25519:
		// ed25519-over-message is Solana today; TON uses 'ton-proof'.
		return VerifySolana(p.Message, p.Signature, p.Address), true
	case SchemeTonProof:
		return VerifyTon(p), true
	case SchemeBIP322:
		return VerifyBitcoin(p), true
	case SchemeSecp256k1XRPL, SchemeEd25519XRPL:
		return VerifyXrp(p), true
	default:
		return false, false
	}
}

// VerifyProof is the one server-side entry point. Chain-agnostic: parse the
// CAIP-122 message, enforce address/domain/nonce/time, then dispatch to the
// per-chain cryptographic verifier. Fails closed: any unknown scheme or
// malformed input returns {OK: false, Reason}, never panics. Mirrors the TS
// verifyProof exactly, including check order and reasons.
func VerifyProof(proof Proof, expected Expectation) Result {
	parsed, err := ParseSiwxMessage(proof.Message)
	if err != nil {
		return fail(ReasonMalformedMessage)
	}

	// 1. Binding: the signer in the message must match the proof's address.
	if !addressesEqual(proof.Chain, parsed.Address, proof.Address) {
		return fail(ReasonAddressMismatch)
	}
	if expected.Address != "" && !addressesEqual(proof.Chain, proof.Address, expected.Address) {
		return fail(ReasonAddressMismatch)
	}

	// 2. Domain + nonce binding (anti-phishing, anti-replay).
	if parsed.Domain != expected.Domain {
		return fail(ReasonDomainMismatch)
	}
	if parsed.Nonce != expected.Nonce {
		return fail(ReasonNonceMismatch)
	}

	// 3. Time window.
	now := expected.Now
	if now.IsZero() {
		now = time.Now()
	}
	skew := expected.ClockSkew
	if skew == 0 {
		skew = DefaultClockSkew
	}
	if exp, ok := parseTime(parsed.ExpirationTime); ok {
		// now > exp + skew
		if now.After(exp.Add(skew)) {
			return fail(ReasonExpired)
		}
	}
	if nbf, ok := parseTime(parsed.NotBefore); ok {
		// now + skew < nbf
		if now.Add(skew).Before(nbf) {
			return fail(ReasonNotYetValid)
		}
	}
	if iat, ok := parseTime(&parsed.IssuedAt); ok {
		// iat - skew > now
		if iat.Add(-skew).After(now) {
			return fail(ReasonNotYetValid)
		}
	}

	// 4. Cryptographic signature.
	ok, supported := crypto(proof)
	if !supported {
		return fail(ReasonUnsupportedScheme)
	}
	if !ok {
		return fail(ReasonBadSignature)
	}

	return Result{OK: true, Address: proof.Address, Chain: proof.Chain}
}
